src/cocotb/share/lib/vpi/VpiImpl.cpp

```diff
@@ -739,6 +744,25 @@ int32_t handle_vpi_callback(p_cb_data cb_data) {
     return rv;
 }
 
+// Main re-entry point for callbacks from simulator
+int32_t handle_vpi_callback(p_cb_data cb_data) {
```

Comment:
If we ever want simulator vendors to solve this on their side we need to provide a way for them to test it. Can you introduce a compile-time define to disable the queuing? (Or alternatively, we give up on requiring simulators to be strictly single-threaded and close the issues.)